recipes/sharedtags-greedyview.lua

```diff
@@ -0,0 +1,195 @@
+-- @treturn bool Whether the tag was moved.
+function sharedtags.movetag(tag, screen)
+    screen = screen or awful.screen.focused()
+    local oldscreen = tag.screen
+
+    -- If the specified tag is allocated to another screen, we need to move it.
+    if oldscreen ~= screen then
+        -- greedyview xmonad-style
+        if tag.selected then
+            local newsel = screen.selected_tag
+            newsel.screen = oldscreen
+        end
+        local oldsel = oldscreen.selected_tag
+        tag.screen = screen
+
+        if oldsel == tag then
+            -- The tag has been moved away. In most cases the tag history
+            -- function will find the best match, but if we really want we can
+            -- try to find a fallback tag as well.
+            if not oldscreen.selected_tag then
+                local newtag = awful.tag.find_fallback(oldscreen)
+                if newtag then
+                    newtag:view_only()
+                end
+            end
+        --else
+            -- NOTE: A bug in awesome 4.0 is causing all tags to be deselected
+            -- here. A shame, but I haven't found a nice way to work around it
+            -- except by fixing the bug (history seems to be in a weird state).
+        end
+
+        -- Also sort the tag in the taglist, by reapplying the index. This is just a nicety.
+        local unpack = unpack or table.unpack
+        for _,s in ipairs({ screen, oldscreen }) do
+            local tags = { unpack(s.tags) } -- Copy
+            table.sort(tags, function(a, b) return a.sharedtagindex < b.sharedtagindex end)
+            for i,t in ipairs(tags) do
+                t.index = i
+            end
+        end
+
+        return true
+    end
+
+    return false
+end
+
+--- View the specified tag on the specified screen.
+-- @param tag The only tag to view.
+-- @tparam[opt=awful.screen.focused()] number screen The screen to view the tag on.
+function sharedtags.viewonly(tag, screen)
+    sharedtags.movetag(tag, screen)
+    tag:view_only()
+end
+
+--- Toggle the specified tag on the specified screen.
+-- The tag will be selected if the screen changes, and toggled if it does not
+-- change the screen.
+-- @param tag The tag to toggle.
+-- @tparam[opt=awful.screen.focused()] number screen The screen to toggle the tag on.
+function sharedtags.viewtoggle(tag, screen)
+    local oldscreen = tag.screen
+
+    if sharedtags.movetag(tag, screen) then
+        -- Always mark the tag selected if the screen changed. Just feels a lot
+        -- more natural.
+        tag.selected = true
+        -- Update the history on the old and new screens.
+        oldscreen:emit_signal("tag::history::update")
+        tag.screen:emit_signal("tag::history::update")
+    else
+        -- Only toggle the tag unless the screen moved.
+        awful.tag.viewtoggle(tag)
+    end
+end
+
+return setmetatable(sharedtags, { __call = function(...) return sharedtags.new(select(2, ...)) end })
```

Comment:
Why the `select` trick instead of just `__call = function(self, args) return sharedtags.new(args) end` (which is IMO more readable).